README.md
# koa-websocket

[![Circle CI](https://circleci.com/gh/kudos/koa-websocket.svg?style=svg)](https://circleci.com/gh/kudos/koa-websocket)

> Koa v2 is now the default. For Koa v1 support install with koa-websocket@2 and see the `legacy` branch.

Supports `ws://` and `wss://`

## Installation

`npm install koa-websocket`

## Usage

```js
const Koa = require('koa'),
  route = require('koa-route'),
  websockify = require('koa-websocket');

const app = websockify(new Koa());

// Regular middleware
// Note it's app.ws.use and not app.use
app.ws.use(function(ctx, next) {
  // return `next` to pass the context (ctx) on to the next ws middleware
  return next(ctx);
});

// Using routes
app.ws.use(route.all('/test/:id', function (ctx) {
  // `ctx` is the regular koa context created from the `ws` onConnection `socket.upgradeReq` object.
  // the websocket is added to the context on `ctx.websocket`.
  ctx.websocket.send('Hello World');
  ctx.websocket.on('message', function(message) {
    // do something with the message from client
        console.log(message);
  });
}));

app.listen(3000);
```

Example with Let's Encrypt ([the Greenlock package](https://git.daplie.com/Daplie/greenlock-koa)):

```js
const Koa = require('koa');
const greenlock = require('greenlock-express');
const websockify = require('koa-websocket');
 
const le = greenlock.create({
  // all your sweet Let's Encrypt options here
});
 
// the magic happens right here
const app = websockify(new Koa(), wsOptions, le.httpsOptions);
 
app.ws.use((ctx) => {
   // the websocket is added to the context as `ctx.websocket`.
  ctx.websocket.on('message', function(message) {
    // do something
  });
});
 
app.listen(3000);
```

With custom websocket options.

```js
const Koa = require('koa'),
  route = require('koa-route'),
  websockify = require('koa-websocket');

const wsOptions = {};
const app = websockify(new Koa(), wsOptions);

app.ws.use(route.all('/', (ctx) => {
   // the websocket is added to the context as `ctx.websocket`.
  ctx.websocket.on('message', function(message) {
    // print message from the client
    console.log(message);
  });
}));

app.listen(3000);
```

## API
#### websockify(KoaApp, [WebSocketOptions], [httpsOptions])
The WebSocket options object just get passed right through to the `new WebSocketServer(options)` call.

The optional HTTPS options object gets passed right into `https.createServer(options)`. If the HTTPS options are 
passed in, koa-websocket will use the built-in Node HTTPS server to provide support for the `wss://` protocol.

## License
MIT


CHANGELOG.md
# Changelog

Note: koa-websocket follows semver.

## v6.0.0

* Upgrade ws to 7.0.1, which drops support for Node 6.
* Upgrade dev deps to fix a vuln with js-yaml.

## v5.0.1

* Remove accidental addition of npm from dependencies

## v5.0.0

* Upgraded all dependencies, including for ws to fix some security issues.

## v4.1.0

* Added support for passing https server options, enabling wss support - @stripedpajamas

## v4.0.0

* Switch Koa v2 support to master, and moved v1 to the legacy branch.
* Added ability to set websocket options.
* Upgraded to websocket 2.x.

## v3.0.1

Because NPM wouldn't let me retroactively tag 3.0.0 as `next` instead of latest.

## v3.0.0

This is a pre-release for compatibility with Koa 2. Doing `npm install koa-websocket` will continue to give you 2.x until Koa 2 is stable.

### Improvements

* Koa 2 comptiblity.

## v2.0.0

### Improvements

* Bump ws major for DoS fix.

## v1.1.0

### Improvements

* Bump ws for iojs 3.

## v1.0.0

### Improvements

* Expose koa context object instead of socket as `this`, attach active ws socket to context object on `websocket`.

### Bugs

* Fix ws urls containing query parameters.


test/test.js
const assert = require('assert');
const fs = require('fs');
const WebSocket = require('ws');
const Koa = require('koa');
const route = require('koa-route');

const koaws = require('..');

describe('should route ws messages seperately', () => {
  const app = koaws(new Koa());

  app.ws.use((ctx, next) => {
    ctx.websocket.on('message', (message) => {
      if (message === '123') {
        ctx.websocket.send(message);
      }
    });
    return next();
  });

  app.ws.use(route.all('/abc', (ctx) => {
    ctx.websocket.on('message', (message) => {
      ctx.websocket.send(message);
    });
  }));

  app.ws.use(route.all('/abc', (ctx) => {
    ctx.websocket.on('message', (message) => {
      ctx.websocket.send(message);
    });
  }));

  app.ws.use(route.all('/def', (ctx) => {
    ctx.websocket.on('message', (message) => {
      ctx.websocket.send(message);
    });
  }));

  let server = null;

  before((done) => {
    server = app.listen(done);
  });

  after((done) => {
    server.close(done);
  });

  it('sends 123 message to any route', (done) => {
    const ws = new WebSocket(`ws://localhost:${server.address().port}/not-a-route`);
    ws.on('open', () => {
      ws.send('123');
    });
    ws.on('message', (message) => {
      assert(message === '123');
      done();
      ws.close();
    });
  });

  it('sends abc message to abc route', (done) => {
    const ws = new WebSocket(`ws://localhost:${server.address().port}/abc`);
    ws.on('open', () => {
      ws.send('abc');
    });
    ws.on('message', (message) => {
      assert(message === 'abc');
      done();
      ws.close();
    });
  });

  it('sends def message to def route', (done) => {
    const ws = new WebSocket(`ws://localhost:${server.address().port}/def`);
    ws.on('open', () => {
      ws.send('def');
    });
    ws.on('message', (message) => {
      assert(message === 'def');
      done();
      ws.close();
    });
  });

  it('handles urls with query parameters', (done) => {
    const ws = new WebSocket(`ws://localhost:${server.address().port}/abc?foo=bar`);
    ws.on('open', () => {
      ws.send('abc');
    });
    ws.on('message', (message) => {
      assert(message === 'abc');
      done();
      ws.close();
    });
  });
});

describe('should support custom ws server options', () => {
  const app = koaws(new Koa(), {
    handleProtocols(protocols) {
      if (protocols.indexOf('bad_protocol') !== -1) { return false; }
      return protocols.pop();
    },
  });

  let server = null;

  before((done) => {
    server = app.listen(done);
  });

  after((done) => {
    server.close(done);
  });

  it('reject bad protocol use wsOptions', (done) => {
    const ws = new WebSocket(`ws://localhost:${server.address().port}/abc`, ['bad_protocol']);
    ws.on('error', () => {
      assert(true);
      done();
      ws.close();
    });
  });
});

describe('should support custom http server options', () => {
  // The cert is self-signed.
  process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

  const secureApp = koaws(new Koa(), {}, {
    key: fs.readFileSync('./test/key.pem'),
    cert: fs.readFileSync('./test/cert.pem'),
  });

  let server = null;

  before((done) => {
    server = secureApp.listen(done);
  });

  after((done) => {
    server.close(done);
  });

  it('supports wss protocol', (done) => {
    const ws = new WebSocket(`wss://localhost:${server.address().port}/abc`);
    ws.on('open', () => {
      assert(true);
      done();
      ws.close();
    });
  });
});


index.js
const url = require('url');
const https = require('https');
const compose = require('koa-compose');
const co = require('co');
const ws = require('ws');

const WebSocketServer = ws.Server;
const debug = require('debug')('koa:websockets');

function KoaWebSocketServer(app) {
  this.app = app;
  this.middleware = [];
}

KoaWebSocketServer.prototype.listen = function listen(options) {
  this.server = new WebSocketServer(options);
  this.server.on('connection', this.onConnection.bind(this));
};

KoaWebSocketServer.prototype.onConnection = function onConnection(socket, req) {
  debug('Connection received');
  socket.on('error', (err) => {
    debug('Error occurred:', err);
  });
  const fn = co.wrap(compose(this.middleware));

  const context = this.app.createContext(req);
  context.websocket = socket;
  context.path = url.parse(req.url).pathname;

  fn(context).catch((err) => {
    debug(err);
  });
};

KoaWebSocketServer.prototype.use = function use(fn) {
  this.middleware.push(fn);
  return this;
};

module.exports = function middleware(app, wsOptions, httpsOptions) {
  const oldListen = app.listen;
  app.listen = function listen(...args) {
    debug('Attaching server...');
    if (typeof httpsOptions === 'object') {
      const httpsServer = https.createServer(httpsOptions, app.callback());
      app.server = httpsServer.listen(...args);
    } else {
      app.server = oldListen.apply(app, args);
    }
    const options = { server: app.server };
    if (wsOptions) {
      Object.keys(wsOptions).forEach((key) => {
        if (Object.prototype.hasOwnProperty.call(wsOptions, key)) {
          options[key] = wsOptions[key];
        }
      });
    }
    app.ws.listen(options);
    return app.server;
  };
  app.ws = new KoaWebSocketServer(app);
  return app;
};


examples/http-and-ws.js
/*
 * Serve HTTP & WebSockets in same application
 */

'use strict';

const Koa        = require('koa'),
      Router     = require('koa-router'),
      webSockify = require('../'),
      app        = new Koa(),
      router     = new Router();

// HTTP routes
router.get('/http', ctx => {
    if (ctx.websocket)
        ctx.throw(400, `ERROR: ${ctx.url} does not handle WebSocket requests`);
    ctx.body = 'Hello World';
});

// WebSocket routes
router.get('/websocket', ctx => {
    if (!ctx.websocket)
        ctx.throw(400, `ERROR: ${ctx.url} only handles WebSocket requests`);
    ctx.websocket.send('Hello World');
    ctx.websocket.on('message', message => {
        console.log(message);
    });
    delete ctx.websocket;
});

app.use(router.routes())
   .use(router.allowedMethods());

// reject unimplemented routes
app.use(ctx => {
    ctx.throw(404, `ERROR: ${ctx.url} not implemented`);
});

// start server
webSockify(app)._webSocketsListen(app.listen(3000));


examples/attach-to-server.js
/*
 * Don't use app.listen(), instead attach to existing server
 */

'use strict';

const Koa        = require('koa'),
      Router     = require('koa-router'),
      http       = require('http'),  // or 'https'
      webSockify = require('../'),
      app        = new Koa(),
      router     = new Router(),
      server     = http.createServer(app.callback());

// WebSocket routes
router.get('/*', ctx => {
    if (!ctx.websocket)
        ctx.throw(400, `ERROR: ${ctx.url} only handles WebSocket requests`);
    ctx.websocket.send('Hello World');
    ctx.websocket.on('message', message => {
        console.log(message);
    });
    delete ctx.websocket;
});

app.use(router.routes())
   .use(router.allowedMethods());

// attach WebSockets to existing server
webSockify(app)._webSocketsListen(server);

// start server
server.listen(3000);
